scipy/spatial/src/distance_pybind.cpp

```diff
@@ -212,108 +367,983 @@ void cdist_weighted_impl(ArrayDescriptor out, T* out_data,
     }
 }
 
-// Extract shape and stride information from NumPy array. Converts byte-strides
-// to element strides, and avoids an extra pointer indirection on access.
-ArrayDescriptor get_descriptor(const py::array& arr) {
-    const auto ndim = arr.ndim();
-    ArrayDescriptor desc(ndim);
+template <typename T>
+void cosine_cdist_weighted_impl(ArrayDescriptor out, T* out_data,
+                         ArrayDescriptor x, const T* x_data,
+                         ArrayDescriptor y, const T* y_data,
+                         ArrayDescriptor w, const T* w_data,
+                         CosineWeightedDistanceFunc<T> f) {
 
-    const auto arr_shape = arr.shape();
-    desc.shape.assign(arr_shape, arr_shape + ndim);
+    const auto num_rowsX = x.shape[0];
+    const auto num_rowsY = y.shape[0];
+    const auto num_cols = x.shape[1];
 
-    desc.element_size = arr.itemsize();
-    const auto arr_strides = arr.strides();
-    desc.strides.assign(arr_strides, arr_strides + ndim);
-    for (intptr_t i = 0; i < ndim; ++i) {
-        if (arr_shape[i] <= 1) {
-            // Under NumPy's relaxed stride checking, dimensions with
-            // 1 or fewer elements are ignored.
-            desc.strides[i] = 0;
-            continue;
```

Comment:
The diff is a bit hard to read because of its size--did you actually delete any of Peter's old infrastructure, or is just getting relocated because of the number of changes here?